subsys/bluetooth/services/mds.c

```diff
@@ -0,0 +1,536 @@
+	k_work_schedule(&mds_instance.mds_work, K_NO_WAIT);
+}
+
+static void stream_disable(struct bt_conn *conn)
+{
+	bool enabled;
+
+	if (conn && (conn != mds_instance.conn)) {
+		return;
+	}
+
+	enabled = atomic_test_and_clear_bit(&mds_instance.stream_state, STREAM_ENABLED);
+	if (enabled) {
+		k_work_reschedule(&mds_instance.mds_work, K_NO_WAIT);
+	}
+}
+
+static bool stream_mode_handler(struct bt_conn *conn, enum data_export_mode stream_mode)
+{
+	bool valid = true;
+
+	switch (stream_mode) {
+	case DATA_EXPORT_MODE_STREAMING_ENABLE:
+		stream_enable(conn);
+		break;
+
+	case DATA_EXPORT_MODE_STREAMING_DISABLE:
+		stream_disable(conn);
+		break;
+
+	default:
+		valid = false;
+		break;
+	}
+
+	return valid;
+}
+
+static ssize_t data_export_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				 const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	enum data_export_mode mode;
+
+	LOG_DBG("MDS Data Export characteristic write, handle %d, conn: %p", attr->handle,
+		(void *)conn);
+
+	if (!bt_gatt_is_subscribed(conn, attr, BT_GATT_CCC_NOTIFY)) {
+		LOG_DBG("MDS Data Export notification are disabled");
+		return BT_GATT_ERR(MDS_ATT_ERROR_CLIENT_NOT_SUBSCRIBED);
+	}
+
+	if (mds_instance.cb->access_enable) {
+		if (!mds_instance.cb->access_enable(conn)) {
+			return BT_GATT_ERR(BT_ATT_ERR_WRITE_NOT_PERMITTED);
+		}
+	}
+
+	if (offset != 0) {
+		LOG_WRN("MDS Data Export invalid write offset");
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len > sizeof(uint8_t)) {
+		LOG_WRN("MDS Data Export invalid attribute length");
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	mode = (enum data_export_mode)(((uint8_t *)buf)[0]);
+
+	if (!stream_mode_handler(conn, mode)) {
+		LOG_WRN("MDS Data Export characteristic write invalid value");
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+
+	return len;
+}
+
+static void data_export_ccc_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	LOG_DBG("MDS Data Export CCCD changed, handle: %u, value: 0x%04X",
+		attr->handle, value);
+}
+
+static void conn_subscription_check(struct bt_conn *conn, void *user_data)
+{
+	struct mds_subscription *mds_subscription = user_data;
+
+	if (bt_gatt_is_subscribed(conn, mds_subscription->attr, BT_GATT_CCC_NOTIFY)) {
+		mds_subscription->conn = conn;
+		mds_subscription->subscribed = true;
+	}
+}
+
+static ssize_t data_export_ccc_write(struct bt_conn *conn,
+				     const struct bt_gatt_attr *attr, uint16_t value)
+{
+	struct mds_subscription mds_subscription = {
+		.attr = attr
+	};
+
+	if ((value != BT_GATT_CCC_NOTIFY) && (value != 0)) {
+		return BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+	}
+
+	if (mds_instance.cb->access_enable) {
+		if (!mds_instance.cb->access_enable(conn)) {
+			return BT_GATT_ERR(BT_ATT_ERR_WRITE_NOT_PERMITTED);
+		}
+	}
+
+	/* Allow only one subscription to the Memfault Data Export characteristic. */
+	bt_conn_foreach(BT_CONN_TYPE_LE, conn_subscription_check, &mds_subscription);
```

Comment:
Maybe we should make our internal check before asking application? Otherwise application may provide input which has no effect.

Reply:
Yes, good idea